test/verify/run-tests

```diff
@@ -21,32 +82,124 @@ def check_valid(filename):
+                    continue
+                if getattr(test_method, "_testlib__non_destructive", False):
+                    serial_tests.append(Test(test_id, build_command(filename, test_str, opts)))
+                else:
+                    parallel_tests.append(Test(test_id, build_command(filename, test_str, opts)))
+                test_id += 1
+
+    print("1..{0}".format(len(parallel_tests) + len(serial_tests)))
+    sys.stdout.flush()
+
+    if serial_tests:
+        testlib.MachineCase.get_global_machine()
+    for test in serial_tests:
+        test.command.insert(-2, "--machine")
+        test.command.insert(-2, "{0}:{1}".format(testlib.MachineCase.get_global_machine().ssh_address,
+                                                 testlib.MachineCase.get_global_machine().ssh_port))
+        test.command.insert(-2, "--browser")
+        test.command.insert(-2, "{0}:{1}".format(testlib.MachineCase.get_global_machine().web_address,
+                                                 testlib.MachineCase.get_global_machine().web_port))
+
+    running_tests = []
+    serial_test = None
+    while serial_tests or parallel_tests or running_tests:
+        if len(running_tests) < jobs:
+            if serial_tests and serial_test is None:
+                test = serial_tests.pop(0)
+                test.process = subprocess.Popen(test.command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+                fcntl.fcntl(test.process.stdout.fileno(), fcntl.F_SETFL, os.O_NONBLOCK)
+                running_tests.append(test)
+                serial_test = test
+            elif parallel_tests:
+                test = parallel_tests.pop(0)
+                test.process = subprocess.Popen(test.command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+                fcntl.fcntl(test.process.stdout.fileno(), fcntl.F_SETFL, os.O_NONBLOCK)
+                running_tests.append(test)
+
+        for test in running_tests.copy():
+            if test.process.poll() is not None:
+                running_tests.remove(test)
+                retry, test_result = finish_test(opts, test)
+                result += test_result
+
+                # run again if needed
+                if retry and test is serial_test:
+                    serial_tests.insert(0, test)
+                    test.output = b""
+                elif retry:
+                    parallel_tests.insert(0, test)
+                    test.output = b""
+                if test is serial_test:
+                    serial_test = None
+
+            else:
+                output = test.process.stdout.read()
+                if output:
+                    test.output += output
+
+        # Wait if we're at maximum number of threads or waiting for the serial_tests to finish
+        if len(running_tests) == jobs or (not parallel_tests and serial_tests):
+            for test in running_tests:
+                try:
+                    test.process.wait(1 / jobs)
```

Comment:
This feels a little excessively often and IMHO is scaling the wrong way around -- with more jobs it would poll more often, and cause run-tests to wake up and go through the loop a lot. I think it's better to use a a `None` timeout for jobs == 1 (no need to poll), and maybe 1s for jobs == 1 and 5s for jobs > 1 seems fine.

However, let's remove the guesswork and polling and just let the kernel tell us when a child finished -- [sigwait()](https://docs.python.org/3/library/signal.html#signal.sigwait) (or `pause()`, but that requires installing a signal handler for SIGCHLD) is meant to do exactly this, so you could just wait for a `SIGCHLD`.

Reply:
Simple example:

```python
#!/usr/bin/python3

import signal
import subprocess

signal.pthread_sigmask(signal.SIG_BLOCK, [signal.SIGCHLD])

pa = subprocess.Popen(['sleep', '5'])
print("started process A, pid", pa.pid)
pb = subprocess.Popen(['sleep', '3'])
print("started process B, pid", pb.pid)

r = signal.sigwait([signal.SIGCHLD])
print("sigwait 1 finished with", r)
# should be process b
assert pb.poll() is not None

r = signal.sigwait([signal.SIGCHLD])
print("sigwait 2 finished with", r)
# should be process a
assert pa.poll() is not None
```
There's also `sigwaitinfo()` which can tell you which pid returned, but I suppose for simplicity it's fine to just go through the loop when any child finished.

Reply:
This would almost certainly lead to a deadlock: we'd stop reading from the `stdout` pipes of the tests, resulting in them blocking on `write()` due to a full pipe buffer and never exiting.

My preferred approach would be to get rid of the non-blocking IO (and excessive polling-via-iteration) above and replace it with a `poll()` on all of the stdout pipes of the tests.  That would also let us detect the EOF case, which would automatically let us know when tests terminate.

Of course this exposes us to the leaked-fd issue......

Reply:
Honestly, it starts to feel like this could all be replaced with a pretty simple GLib-using C program..... it's not that C is a better language for this sort of thing than Python, but rather that this sort of programming benefits greatly from a mainloop.... and unless we can add a dep on twisted, the next available mainloop that we already depend on is in GLib....

Could also use GLib APIs from Python, I guess... but then I guess we also don't depend on the python bindings?

Let's chat about this on IRC.

Reply:
Right, sorry for the bad advice -- this of course needs to keep looping to slurp the subprocesses' stdout. So aside from some minor bikeshedding/optimization this is basically fine, and should not block this PR.

Reply:
I just had another idea: As we don't need live streaming anyway, you could also have subprocesses output their stuff into a temporary file (could even be an anonymous one), and copy that to stdout once the process exits. That would simplify the loop a lot, and you can again use the SIGCHLD approach to avoid useless waiting and polling.

Reply:
We all know that I like a `memfd` as much as the next hacker... :)

We could also use a thread pool and do a simple subprocess.check_output() from each thread...

I guess probably Sanne doesn't feel like rewriting it again, though.

Reply:
Full ack, let's keep it as-is for this PR. We can go wild in a follow-up.